pkg/stores/proxy/cache/cache.go

```diff
@@ -0,0 +1,150 @@
+	size int
+	obj  interface{}
+}
+
+// NewSizedRevisionCache accepts a sizeLimit, in bytes, and a maxElements parameter to create a SizedRevisionCache.
+func NewSizedRevisionCache(sizeLimit, maxElements int) *SizedRevisionCache {
+	return &SizedRevisionCache{
+		listRevisionCache: cache.NewLRUExpireCache(maxElements),
+		sizeLimit:         sizeLimit,
+	}
+}
+
+// String returns a string contains the fields values of the cacheKey receiver.
+func (c CacheKey) String() string {
+	return fmt.Sprintf("listOptions: %v, resourcePath: %s, namespace: %s, cont: %s", c.listOptions.String(), c.resourcePath, c.namespace, c.cont)
+}
+
+// Get returns the UnstructuredList stored under the given cacheKey if available. If not, as error is returned.
+func (s *SizedRevisionCache) Get(key CacheKey) (*unstructured.UnstructuredList, error) {
+	s.cacheLock.Lock()
+	defer s.cacheLock.Unlock()
+	// check if cache stored all namespaces
+
+	obj, ok := s.listRevisionCache.Get(key)
+	if !ok {
+		return nil, ErrNotFound
+	}
+	uList, ok := obj.(cacheObj)
+	if !ok {
+		return nil, fmt.Errorf("could not assert object stored with key [%s] key as UnstructuredList", key)
+	}
+	return uList.obj.(*unstructured.UnstructuredList), nil
+}
+
+// Add attempts to add the given UnstructuredListed under the given key.
+func (s *SizedRevisionCache) Add(key CacheKey, list *unstructured.UnstructuredList) error {
+	s.cacheLock.Lock()
+	defer s.cacheLock.Unlock()
+
+	objBytes, err := list.MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	cacheListObj := cacheObj{
+		size: len(objBytes),
+		obj:  list,
+	}
+
+	currentSize := s.sizeOfCurrentEntry(key)
+
+	if err = s.adjustSize(cacheListObj.size - currentSize); err != nil {
+		return err
+	}
+	s.listRevisionCache.Add(key, cacheListObj, 30*time.Minute)
+
+	return nil
+}
+
+func (s *SizedRevisionCache) calculateSize() int {
+	var total int
+	for _, key := range s.listRevisionCache.Keys() {
+		total += s.sizeOfCurrentEntry(key)
+	}
+	return total
+}
+
+func (s *SizedRevisionCache) sizeOfCurrentEntry(key interface{}) int {
+	obj, ok := s.listRevisionCache.Get(key)
+	if !ok {
+		return 0
+	}
+
+	cacheObject, ok := obj.(cacheObj)
+	if !ok {
+		return 0
+	}
+
+	return cacheObject.size
+}
+
+func (s *SizedRevisionCache) adjustSize(diff int) error {
+	if !(s.size+diff > s.sizeLimit) {
+		s.size += diff
+		return nil
+	}
+
+	// the size is recalculated here to check whether entries have expired that would
+	// make s.size+diff stay below sizeLimit.
+	s.size = s.calculateSize()
+
+	if !(s.size+diff > s.sizeLimit) {
+		s.sizeLimit += diff
+		return nil
+	}
+
+	return fmt.Errorf("[steve proxy cache]: cache is near full with a size of [%d] and limit of [%d], cannot increment size by [%d]", s.size, s.sizeLimit, diff)
+}
+
+// GetCacheKey returns a cacheKey with the given field values set.
+func GetCacheKey(options v1.ListOptions, resourcePath, ns, cont string) CacheKey {
+	return CacheKey{
+		listOptions:  options,
+		resourcePath: resourcePath,
```

Comment:
Is it correct to say that `resourcePath` is set given `options.Kind` and `options.APIVersion`?

If so:

Nitpick: it could be removed from CacheKey.

Reply:
`ListOptions.TypeMeta.Kind` and `ListOptions.TypeMeta.APIVersion` pertain to the `ListOptions` type, not to the actual requested resource, so they should not be used in place of `resourcePath`.

Reply:
That is correct. ListOptions can be, and often will be aside from resourceVersion, empty.

Reply:
Thanks for explaining